evap/contributor/views.py

```diff
@@ -21,22 +23,40 @@ def index(request):
     show_delegated = get_parameter_from_url_or_session(request, "show_delegated", True)
 
     contributor_visible_states = ['prepared', 'editor_approved', 'approved', 'in_evaluation', 'evaluated', 'reviewed', 'published']
-    own_evaluations = Evaluation.objects.filter(contributions__contributor=user, state__in=contributor_visible_states)
+    own_courses = Course.objects.filter(
+        Q(evaluations__state__in=contributor_visible_states) & (
+            Q(responsibles=user) |
+            Q(evaluations__contributions__contributor=user)
+        )
+    )
+    own_evaluations = [evaluation for course in own_courses for evaluation in course.evaluations.all() if evaluation.state != "new" and evaluation.can_user_see_evaluation(user)]
+    for evaluation in own_evaluations:
+        evaluation.contributes_to = evaluation.contributions.filter(contributor=user).exists()
 
-    displayed_evaluations = list(own_evaluations)
+    displayed_evaluations = set(own_evaluations)
     if show_delegated:
         represented_users = user.represented_users.all()
-        delegated_evaluations = Evaluation.objects.exclude(id__in=own_evaluations).filter(contributions__can_edit=True, contributions__contributor__in=represented_users, state__in=contributor_visible_states)
+        delegated_courses = Course.objects.filter(
+            Q(evaluations__state__in=contributor_visible_states) & (
+                Q(responsibles__in=represented_users) |
+                Q(evaluations__contributions__can_edit=True, evaluations__contributions__contributor__in=represented_users)
+            )
+        )
+        delegated_evaluations = [evaluation for course in delegated_courses for evaluation in course.evaluations.all() if evaluation.state != "new" and evaluation.can_user_see_evaluation(user)]
         for evaluation in delegated_evaluations:
             evaluation.delegated_evaluation = True
-        displayed_evaluations += list(delegated_evaluations)
-    displayed_evaluations.sort(key=lambda evaluation: list(STATES_ORDERED.keys()).index(evaluation.state))
-
-    delegate_selection_form = DelegateSelectionForm()
+        displayed_evaluations.update(delegated_evaluations)
+    displayed_evaluations = list(displayed_evaluations)
+    displayed_evaluations.sort(key=lambda evaluation: evaluation.full_name)  # evaluations must be sorted for regrouping them in the template
 
     for evaluation in displayed_evaluations:
-        evaluation.distribution = calculate_average_distribution(evaluation)
-        evaluation.avg_grade = distribution_to_grade(evaluation.distribution)
+        if evaluation.state == "published":
+            if not evaluation.is_single_result:
+                evaluation.distribution = calculate_average_distribution(evaluation)
```

Comment:
this calculates the results for the evaluations, `get_evaluations_with_course_result_attributes` calculates the courses' results. is this still a problem after addressing what you commented on `get_evaluations_with_course_result_attributes` above?

Reply:
i understood that, i took issue with doing that work multiple times for the same evaluation. as far as i can see, if you restructure `get_evaluations_with_course_result_attributes` to do that work per course, you're still doing it twice per evaluation. see the top comment of the review for some input on how this might be restructured.